### DISCUSSION TITLE
how deep sleep processor, there is still a signal on the interrupt?

### DISCUSSION THREAD

**Discussion by Brendy1986** [SpenceKonde/megaTinyCore#923] [Q&A]
I wake up Attiny3226 with a Falling interrupt. I noticed that when I press and release the button it goes into a deep sleep, but when I press and hold the button I get 120uA.
My question is how to put the tiny into a deep sleep despite the low state on interrupt.
I used the PB4 reset pin to wake up. I added a 10k resistor and a 100nF capacitor and it works fine.
can I use interrupt?

**Comment by SpenceKonde** [comment]
Would need to see code and schematic to comment further. (hand scrawled schematic that is accurate is preferred over a fritzing diagram that has had parts substituted)

 Presumably there's a pullup on the button pin to keep it from floating and triggering randomly. But whenever it is held low, current will flow through the pullup and button, and that could easily account for 120uA. things with pullups being held low are one of the things that you need to avoid when power consumption is a concern.

Also, moved to Q&A section

**Reply by Brendy1986** [reply]
after applying this, although the state persists, it goes to sleep, I have to disconnect and reconnect the power again and the program will execute again.
I wonder if there is any way to use interrupt?
![Screenshot_2023-03-16-23-47-42-737_com android chrome](https://user-images.githubusercontent.com/118827809/225770142-d5e3257d-61f0-42c8-ab65-668fd8e6961c.jpg)

**Comment by Brendy1986** [comment]
If I wake up RISING will I not have 120uA current consumption?
Please tell me how to wake up RISING I tried but I can't.


```
#include <avr/io.h>
#include <math.h>
#include <avr/interrupt.h>
#include <avr/sleep.h>
#include <util/delay.h>

volatile bool box = false;

ISR(PORTA_PORT_vect)
{
  if (bitRead(PORTA.INTFLAGS, 6)) {  

    PORTA.INTFLAGS = PORT_INT6_bm; //reset flags
    box = true;
  }
}
void Wake_UP_PA2() {
Serial.println("Wake Up From PA2");
}

void setup() {

  Serial.begin(9600);

  SLPCTRL.CTRLA = 1 << SLPCTRL_SEN_bp       /* Sleep enable: enabled */
                  | SLPCTRL_SMODE_STDBY_gc; /* Standby Mode */

  ADC0.CTRLA = 1 << ADC_ENABLE_bm & ~(1 << ADC_RUNSTDBY_bm);  
  ADC0.CTRLC = ADC_REFSEL_2500MV_gc;                         
  ADC0.CTRLA = (1 << ADC_ENABLE_bp) & (~(1 << ADC_RUNSTDBY_bp) & ~(1 << ADC_LOWLAT_bp)); 


  PORTA.DIRCLR |= 1 << 6; //ustwienie bitu jako input
  PORTA.PIN6CTRL  = PORT_ISC_FALLING_gc | 1 << PORT_PULLUPEN_bp; 

  CPUINT.CTRLA &= ~CPUINT_LVL0RR_bm; // 
  CPUINT.LVL0PRI = PORTA_PORT_vect_num; 

  for (uint8_t i = 0; i < 8; i++) {    
   *((uint8_t *)&PORTA + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  }

  for (uint8_t i = 0; i < 8; i++) {    
    *((uint8_t *)&PORTB + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  }
  for (uint8_t i = 0; i < 8; i++) {   
    *((uint8_t *)&PORTC + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  }
  Serial.begin(9600);   

  sei();    
}
void loop() {

     if(box){
      box = false;
      Wake_UP_PA2();
      delay(500);
  }
  Serial.println("Sleep");
  delay(100); 
  sleep_cpu();
}
```

**Comment by Brendy1986** [comment]
success works. processor sleeps at 100nA even though there is still a signal on the interrupt

**Reply by SpenceKonde** [reply]
The only tricky thing about waking on RISING or FALLING is that it only works on pins 2 and 6 within each port on the 2-series (as well as older tinyAVRs, mega 0-series, DA, and DB parts). DD and later have removed that restriction - all pins are now "fully async". Fully async pins can wake without a clock on rising or falling, instead of just change and low, and can be triggered during normal operation by a signal shorter than one clock cycle. The downside of this this is that it sometimes these sub-clock-cycle transients are noise, not signal, and you'd prefer it not pick them up. 

No matter what you do, however, you want to make sure that you do not have any floating pins (which are neither driven high or low, nor pulled up) on which the "input buffer" is enabled, as these can significantly increase power consumption (depending on ambient electrical noise/interference)  - the pins act as antennas, and cause the input buffer to keep toggling. and that takes non-zero energy. All unused pins should either be set output, have their pullups turned on, or have the input disabled. Microchip's power consumption tests are done with all pins output and low, and with input disabled. I suspect this is done because that's the combination that minimizes power consmption, thus making maximally flattering specs to put in the datasheet.

**Reply by unhinged-engineer** [reply]
> success works. processor sleeps at 100nA even though there is still a signal on the interrupt

What did you change? what worked? I'm struggling with interrupts, so any successful debugging would be helpful to me.

**Reply by Brendy1986** [reply]
> > sukces działa. procesor śpi przy 100nA, mimo że nadal jest sygnał na przerwaniu
> 
> Co zmieniłeś? co zadziałało? Walczę z przerwaniami, więc wszelkie udane debugowanie byłoby dla mnie pomocne.

I am using wakeup interrupt rising. I disabled internal pullup
I did an external pullup by adding a 10M resistor and anti-interference filter resistor + ceramic capacitor.
it's been working fine for almost a month now with hc12 radio 433mhz chip.

**Comment by SpenceKonde** [comment]
Wait why do you want it sleeping during the interrupt? Isn't that what's supposed to wake it up? 

But again, see what I said above, - only pins 2 and 6 on a port (PA2, PA6, PB2, PB6, etc) are "fully async" and can wake without system clock on rising or falling not just change. 
All other pins, though their interrupts would not wake the chip. This is true of ATmega 0-series, ATtiny 0, 1, and 2, series, AVR DA and AVR DB series parts. On the most recent AVR DD and AVR EA parts (and almost certainly future DU and EB series), all pins are "fully async" and can wake on rising and falling from sleep modes where there is no CLKper, which is one of the two magical functions of fully async pins. 

The other thing "fully async" pins can do is respond to very brief events (shorter than a clock cycle). If this is your signal, wonderful! If it's not, we call it noise - noise rejection is markedly worse and scenarios to demonstrate this can be contrived easily enough. But that's what filters are for. 

Also ignore the whole thing about dead time for fully async pins in the datasheet, it's muddled, wrong, and useless. What they meant to say is: 
* On Fully asynchronous pins, there is always a 2-3 cycle synchronization delay between the interrupt condition being true while the flag is clear, and the flag being set. Nothing to do with interrupts - if you're actually triggering interupts, you're looking for a delay about 1/7th the minimum exec time of an interrupt. All 4 types of interrupt conditions can be detected without a clock
* On other pins, if the state of the pin is synchronized and the interrupt condition is met and flag not set, flag will be set immediately. However only change and low level interrupts can be used to wake without a clock. 
* This is best observed with interrupts off and a tight asm loop that alternately clears and checks the flag while the pin is held low and interrupt is set to low level. with cbi, sbis, rjmp .-6 - this will never exit on an async pin, but it will exit immediately on an non-async pin. ]
* That is an extreme corner case; under ordinary, and even extraordinary conditions, this doesn't matter. "But!" you say, "This means the async pins are slower! No, actually they're about the same speed, you're forgetting that there's 2-3 cycles sync time on all digital pins. 
* I believe the silicon is behaving correctly and the datasheet is incorrect, not the other way around.
